File: chaincode/fabcctv/node/fabcctv.js
/*
# Copyright IBM Corp. All Rights Reserved.
#
# SPDX-License-Identifier: Apache-2.0
*/

'use strict';
const shim = require('fabric-shim');
const util = require('util');

let Chaincode = class {

  // The Init method is called when the Smart Contract 'fabcctv' is instantiated by the blockchain network
  // Best practice is to have any Ledger initialization in separate function -- see initLedger()
  async Init(stub) {
    console.info('=========== Instantiated fabcctv chaincode ===========');
    return shim.success();
  }

  // The Invoke method is called as a result of an application request to run the Smart Contract
  // 'fabcctv'. The calling application program has also specified the particular smart contract
  // function to be called, with arguments
  async Invoke(stub) {
    let ret = stub.getFunctionAndParameters();
    console.info(ret);

    let method = this[ret.fcn];
    if (!method) {
      console.error('no function of name:' + ret.fcn + ' found');
      throw new Error('Received unknown function ' + ret.fcn + ' invocation');
    }
    try {
      let payload = await method(stub, ret.params);
      return shim.success(payload);
    } catch (err) {
      console.log(err);
      return shim.error(err);
    }
  }

  async queryFrame(stub, args) {
    if (args.length != 1) {
      throw new Error('Incorrect number of arguments. Expecting frameKey ex:199502221420');
    }
    let frameKey = args[0];

    let frameAsBytes = await stub.getState(frameKey); // get the frame from chaincode state
    if (!frameAsBytes || frameAsBytes.toString().length <= 0) {
      throw new Error(frameKey + ' does not exist: ');
    }
    console.log(frameAsBytes.toString());
    return frameAsBytes;
  }

  async validateFrame(stub, args) {

    if (args.length != 2) {
      throw new Error('Incorrect number of arguments. Expecting 2');
    }
    let frameKey = args[0];

    let frameAsBytes = await stub.getState(frameKey); // get the frame from chaincode state
    if (!frameAsBytes || frameAsBytes.toString().length <= 0) {
      throw new Error(frameKey + ' does not exist: ');
    }
    console.log(frameAsBytes.toString());

    if (frameAsBytes.toString() != args[1]) {
      return frameAsBytes;
    }

    return shim.success();
  }

  async initLedger(stub, args) {
    console.info('============= START : Initialize Ledger ===========');
    let frames = [];
    frames.push({
      time: "199502221420", // TODO: check the format of date is good or not
      data: "Hash" // TODO: insert sample hash image hash value
    });

    for (let i = 0; i < frames.length; i++) {
      frames[i].docType = 'frame';
      await stub.putState('FRAME' + i, Buffer.from(JSON.stringify(frames[i])));
      console.info('Added <--> ', frames[i]);
    }
    console.info('============= END : Initialize Ledger ===========');
  }

  async createFrame(stub, args) {
    console.info('============= START : Create Frame ===========');
    if (args.length != 3) {
      throw new Error('Incorrect number of arguments. Expecting 3');
    }

    var frame = {
      docType: 'frame',
      key: args[1],
      value: args[2]
    };

    await stub.putState(args[0], Buffer.from(JSON.stringify(frame)));
    console.info('============= END : Create Frame ===========');
  }

  async queryFramesByRange(stub, args) {

    if (args.length != 1) {
      throw new Error('Incorrect number of arguments. Expecting 1');
    }

    let startKey = '000000000000'; // this value should not being changed
    let endKey = args[0];

    let iterator = await stub.getStateByRange(startKey, endKey);

    let allResults = [];
    while (true) {
      let res = await iterator.next();

      if (res.value && res.value.value.toString()) {
        let jsonRes = {};
        console.log(res.value.value.toString('utf8'));

        jsonRes.Key = res.value.key;
        try {
          jsonRes.Record = JSON.parse(res.value.value.toString('utf8'));
        } catch (err) {
          console.log(err);
          jsonRes.Record = res.value.value.toString('utf8');
        }
        allResults.push(jsonRes);
      }
      if (res.done) {
        console.log('end of data');
        await iterator.close();
        console.info(allResults);
        return Buffer.from(JSON.stringify(allResults));
      }
    }
  }
};

shim.start(new Chaincode());

File: chaincode/fabcctv/go/fabcctv.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/*
 * The fabcctv smart contract for documentation topic:
 * Writing Your First Blockchain Application
 */

package main

/* Imports
 * 4 utility libraries for formatting, handling bytes, reading and writing JSON, and string manipulation
 * 2 specific Hyperledger Fabric specific libraries for Smart Contracts
 */
import (
	"bytes"
	"encoding/json"
	"fmt"
	"strconv"

	"github.com/hyperledger/fabric/core/chaincode/shim"
	sc "github.com/hyperledger/fabric/protos/peer"
)

// Define the Smart Contract structure
type SmartContract struct {
}

// Define the frame structure, with 4 properties.  Structure tags are used by encoding/json library
type Frame struct {
	Time  string `json:"time"`
	Data  string `json:"data"`
}

/*
 * The Init method is called when the Smart Contract "fabcctv" is instantiated by the blockchain network
 * Best practice is to have any Ledger initialization in separate function -- see initLedger()
 */
func (s *SmartContract) Init(APIstub shim.ChaincodeStubInterface) sc.Response {
	return shim.Success(nil)
}

/*
 * The Invoke method is called as a result of an application request to run the Smart Contract "fabcctv"
 * The calling application program has also specified the particular smart contract function to be called, with arguments
 */
func (s *SmartContract) Invoke(APIstub shim.ChaincodeStubInterface) sc.Response {

	// Retrieve the requested Smart Contract function and arguments
	function, args := APIstub.GetFunctionAndParameters()
	// Route to the appropriate handler function to interact with the ledger appropriately
	if function == "queryFrame" {
		return s.queryFrame(APIstub, args)
	} else if function == "initLedger" {
		return s.initLedger(APIstub)
	} else if function == "createFrame" {
		return s.createFrame(APIstub, args)
	} else if function == "queryFramesByRange" {
		return s.queryFramesByRange(APIstub, args)
	} else if function == "validateFrame" {
		return s.validateFrame(APIstub, args)
	}

	return shim.Error("Invalid Smart Contract function name.")
}

func (s *SmartContract) queryFrame(APIstub shim.ChaincodeStubInterface, args []string) sc.Response {

	if len(args) != 1 {
		return shim.Error("Incorrect number of arguments. Expecting 1")
	}

	frameAsBytes, _ := APIstub.GetState(args[0])
	return shim.Success(frameAsBytes)
}

func (s *SmartContract) validateFrame(APIstub shim.ChaincodeStubInterface, args []string) sc.Response {

	// TODO: change the return as simple value like true or false

	if len(args) != 2 {
		return shim.Error("Incorrect number of arguments. Expecting 2")
	}

	frameAsBytes, _ := APIstub.GetState(args[0])

	if string(frameAsBytes) != args[1] {
		return shim.Success(frameAsBytes)
	}

	return shim.Success(nil)
}

func (s *SmartContract) initLedger(APIstub shim.ChaincodeStubInterface) sc.Response {
	frames := []Frame{
		Frame{Time: "199502221420", Data: "Hash#1"},
		Frame{Time: "199502221421", Data: "Hash#2"},
		Frame{Time: "199502221422", Data: "Hash#3"},
		Frame{Time: "199502221423", Data: "Hash#4"},
		Frame{Time: "199502221424", Data: "Hash#5"},
		Frame{Time: "199502221425", Data: "Hash#6"},
	}

	i := 0
	for i < len(frames) {
		fmt.Println("i is ", i)
		frameAsBytes, _ := json.Marshal(frames[i])
		APIstub.PutState("FRAME"+strconv.Itoa(i), frameAsBytes)
		fmt.Println("Added", frames[i])
		i = i + 1
	}

	return shim.Success(nil)
}

func (s *SmartContract) createFrame(APIstub shim.ChaincodeStubInterface, args []string) sc.Response {

	if len(args) != 3 {
		return shim.Error("Incorrect number of arguments. Expecting 3")
	}

	var frame = Frame{Time: args[1], Data: args[2]}

	frameAsBytes, _ := json.Marshal(frame)
	APIstub.PutState(args[0], frameAsBytes)

	return shim.Success(frameAsBytes)
}

func (s *SmartContract) queryFramesByRange(APIstub shim.ChaincodeStubInterface, args []string) sc.Response {

	if len(args) != 1 {
		return shim.Error("Incorrect number of arguments. Expecting 1")
	}

	startKey := "000000000000" // this value should not being changed
	endKey := args[0] // TODO: value should be defined

	resultsIterator, err := APIstub.GetStateByRange(startKey, endKey)
	if err != nil {
		return shim.Error(err.Error())
	}
	defer resultsIterator.Close()

	// buffer is a JSON array containing QueryResults
	var buffer bytes.Buffer
	buffer.WriteString("[")

	bArrayMemberAlreadyWritten := false
	for resultsIterator.HasNext() {
		queryResponse, err := resultsIterator.Next()
		if err != nil {
			return shim.Error(err.Error())
		}
		// Add a comma before array members, suppress it for the first array member
		if bArrayMemberAlreadyWritten == true {
			buffer.WriteString(",")
		}
		buffer.WriteString("{\"Key\":")
		buffer.WriteString("\"")
		buffer.WriteString(queryResponse.Key)
		buffer.WriteString("\"")

		buffer.WriteString(", \"Record\":")
		// Record is a JSON object, so we write as-is
		buffer.WriteString(string(queryResponse.Value))
		buffer.WriteString("}")
		bArrayMemberAlreadyWritten = true
	}
	buffer.WriteString("]")

	fmt.Printf("- queryFramesByRange:\n%s\n", buffer.String())

	return shim.Success(buffer.Bytes())
}

// The main function is only relevant in unit test mode. Only included here for completeness.
func main() {

	// Create a new Smart Contract
	err := shim.Start(new(SmartContract))
	if err != nil {
		fmt.Printf("Error creating new Smart Contract: %s", err)
	}
}
